added function to track active users

ArnavS59 opened the pull request:
# Track active users in DB

## :recycle: Current situation & Problem
PR for #17


## :gear: Release Notes 
* Added function `extract_latest_user_interaction` to `data_processor.py` to track active users in the DB.


## :books: Documentation
*Please ensure that you properly document any additions in conformance to [Spezi Documentation Guide](https://github.com/StanfordSpezi/.github/blob/main/DOCUMENTATIONGUIDE.md).*
*You can use this section to describe your solution, but we encourage contributors to document your reasoning and changes using in-line documentation.* 


## :white_check_mark: Testing
*Please ensure that the PR meets the testing requirements set by CodeCov and that new functionality is appropriately tested.*
*This section describes important information about the tests and why some elements might not be testable.*


## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

ArnavS59 committed:
added function to track active users

Vicbi committed:
Merge branch 'main' into main

Vicbi commented on src/spezi_data_pipeline/data_processing/data_processor.py:
```diff
@@ -313,3 +313,24 @@ def select_data_by_dates(  # pylint: disable=unused-variable
+        return
+
+    #First convert column to datetime for comparison
+    flattend_fhir_df['EffectiveDateTime'] = pd.to_datetime(flattend_fhir_df['EffectiveDateTime'], format='%d.%m.%y')
```
Consider replacing the string column name with a constant already declared for all column names in `FHIRDataframe`

Vicbi commented on src/spezi_data_pipeline/data_processing/data_processor.py:
```diff
@@ -313,3 +313,24 @@ def select_data_by_dates(  # pylint: disable=unused-variable
+    flattend_fhir_df['EffectiveDateTime'] = pd.to_datetime(flattend_fhir_df['EffectiveDateTime'], format='%d.%m.%y')
+    #Filter the most recent entry for each userid 
+    most_recent_df=flattend_fhir_df.loc[flattend_fhir_df.groupby('UserId')['EffectiveDateTime'].idxmax()]
+    most_recent_df=most_recent_df[['UserId','EffectiveDateTime']]#select the relevant cols
```
Same comment here for `"UserID"` and `"EffectiveDateTime"`.

Vicbi commented on src/spezi_data_pipeline/data_processing/data_processor.py:
```diff
@@ -313,3 +313,24 @@ def select_data_by_dates(  # pylint: disable=unused-variable
+    most_recent_df=flattend_fhir_df.loc[flattend_fhir_df.groupby('UserId')['EffectiveDateTime'].idxmax()]
+    most_recent_df=most_recent_df[['UserId','EffectiveDateTime']]#select the relevant cols
+    most_recent_df.rename(columns={'EffectiveDateTime': 'LastUserInteraction'}, inplace=True)
+    most_recent_df.to_csv('output.csv')
```
Another suggestion would be to allow the user of the `spezi-data-pipeline` package to set the name of the output file by themselves.

Vicbi commented on src/spezi_data_pipeline/data_processing/data_processor.py:
```diff
@@ -313,3 +313,24 @@ def select_data_by_dates(  # pylint: disable=unused-variable
     )
+
+def extract_latest_user_interaction(  # pylint: disable=unused-variable
+    flattend_fhir_df: pd.DataFrame
```
In the `spezi-data-pipeline` package, the functionalities are built around `FHIRDataFrame` objects. You might want to adjust your function to receive `FHIRDataFrame` as an input argument.

Vicbi reviewed (commented):
Hi @ArnavS59,
Thank you for your contribution! I have a few suggestions:
- Consider using string constants for the column names within your function. This can help with readability and future code maintenance.
- Please take a moment to address the linting errors.
- It would be helpful to add a unit test for your function.

vishnuravi committed:
Merge branch 'main' into main

vishnuravi committed:
Address review feedback for extract_latest_user_interaction

- Accept a FHIRDataFrame as input instead of a raw pandas DataFrame
- Replace string column literals with ColumnNames constants
- Add a configurable output_file_path parameter (no CSV written by default)
- Return the resulting summary DataFrame and avoid mutating the input
- Add a LAST_USER_INTERACTION column name constant
- Add unit tests covering extraction, CSV export, immutability, and empty input
- Fix formatting and linting
